x/bundles/keeper/logic_round_robin.go

```diff
@@ -0,0 +1,259 @@
+// The -1 is necessary to put them at the end of the queue (as the queue is centered around zero).
+// The additional -0.125 are added as a penalty factor.
+func (k Keeper) LoadRoundRobinValidatorSet(ctx sdk.Context, poolId uint64) RoundRobinValidatorSet {
+	vs := RoundRobinValidatorSet{}
+	vs.PoolId = poolId
+	vs.Progress = make(map[string]int64, 0)
+	totalDelegation := int64(0)
+	// Used for calculating the set difference of active validators and existing round-robin set
+	newValidators := make(map[string]bool, 0)
+	// Add all current pool validators to the round-robin set
+	for _, address := range k.stakerKeeper.GetAllStakerAddressesOfPool(ctx, poolId) {
+		delegation := k.delegationKeeper.GetDelegationAmount(ctx, address)
+		if delegation > 0 {
+			// If a validator has no delegation do not add to the round-robin set. Validator is basically non-existent.
+			vs.Validators = append(vs.Validators, RoundRobinValidatorPower{
+				Address: address,
+				Power:   int64(delegation),
+			})
+			vs.Progress[address] = 0
+			totalDelegation += int64(delegation)
+			newValidators[address] = true
+		}
+	}
+
+	// Fetch stored progress
+	roundRobinProgress, _ := k.GetRoundRobinProgress(ctx, poolId)
+	for _, progress := range roundRobinProgress.ProgressList {
+		_, ok := vs.Progress[progress.Address]
+		// If the address is not found it means that the validator left the pool.
+		// Therefore, this entry can be ignored.
+		if ok {
+			vs.Progress[progress.Address] += progress.Progress
+		}
+
+		// Progress already exists, validator can not be new.
+		newValidators[progress.Address] = false
+	}
+
+	for newAddress, isNew := range newValidators {
+		if isNew {
+			// Put new validators at the end of the queue with an additional penalty factor.
+			vs.Progress[newAddress] = sdk.MustNewDecFromStr("-1.125").MulInt64(totalDelegation).TruncateInt64()
+		}
+	}
+
+	vs.normalize()
+	return vs
+}
+
+// SaveRoundRobinValidatorSet saves the current round-robin progress for the given poolId to the KV-Store
+func (k Keeper) SaveRoundRobinValidatorSet(ctx sdk.Context, vs RoundRobinValidatorSet) {
+	roundRobinProgress := types.RoundRobinProgress{
+		PoolId:       vs.PoolId,
+		ProgressList: vs.GetRoundRobinProgress(),
+	}
+	k.SetRoundRobinProgress(ctx, roundRobinProgress)
+}
+
+// GetRoundRobinProgress returns a deterministic (sorted) list of the current round-robin progress.
+// Due to the fact that maps have no order in Go, we must introduce one by ourselves.
+// This is done by sorting all entries by their addresses alphabetically.
+func (vs *RoundRobinValidatorSet) GetRoundRobinProgress() []*types.RoundRobinSingleValidatorProgress {
+	// Convert map to list
+	result := make([]*types.RoundRobinSingleValidatorProgress, 0)
+	for address, progress := range vs.Progress {
+		singleProgress := types.RoundRobinSingleValidatorProgress{
+			Address:  address,
+			Progress: progress,
+		}
+		result = append(result, &singleProgress)
+	}
+	// Sort addresses alphabetically
```

Comment:
Hi, this is only for exporting the state, not while operating on it. One would have to re-sort every time the state changes. However min-max is only called once per round. So with sort + lookup it would be O(nlog(n)) instead of O(n). And as this does not operate on the KV Store, performance is not really an issue. We are talking about microseconds here as this is a for-loop over 50 in-memory entries (so almost nothing).